packaging/debian/pushpin-handler.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" (C) Copyright 2013 Jan Niehusmann <jan@debian.org>,
.\"
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH "PUSHPIN-HANDLER" 1 "November  7, 2013"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.de URL
\\$2 \(laURL: \\$1 \(ra\\$3
..
.if \n[.g] .mso www.tmac
.\"
.SH NAME
pushpin-handler \- handles incoming publish requests (part of pushpin)
.SH SYNOPSIS
.B pushpin-handler
.RI [--verbose] " " --config= CONFIG

.SH DESCRIPTION
This starts
.B pushpin-handler
using the specified configuration (pushpin.conf).

In the default configuration, pushpin-handler is started automatically by pushpin.

.SH "SEE ALSO"
\&\fIpushpin\fR\|(1)


CHANGELOG.md
Pushpin Changelog
=================

v. 1.41.0 (2025-08-08)

  * Add support for gone links.
  * WebSocket-Over-HTTP: allow keeping back data (Content-Bytes-Accepted).
  * Support IPv6 addresses in route targets.
  * Cache root certs when making outbound connections.
  * New config option: update_on_first_subscription.
  * New route condition option: log_level.
  * Correctly read push_in_http_max_{headers,body}_size options.
  * Apply next link response's own filters to its content.
  * runner: look for bins in target dir (makes "cargo run" work).
  * Remove use of Qt event loop, to reduce dependence on Qt.

v. 1.40.1 (2024-08-13)

  * Fix build with recent rustc.

v. 1.40.0 (2024-07-17)

  * Restore support for receiving chunked responses from backends.
  * For early errors such as malformed requests, respond instead of disconnecting.
  * Rename Condure to connmgr.
  * Fix crash under high load.
  * Fix hang when receiving large WebSocket rejections from the backend.

v. 1.39.1 (2024-03-18)

  * Regenerate pushpin.conf.inst post-build to ensure up-to-date configuration.
  * Update legacy runner use revised Qt linking logic, aligning with main branch improvements.

v. 1.39.0 (2024-03-14)

  * Add support for multiple proxy worker threads.
  * New config option: workers (under [proxy]).
  * Fix memory leak when proxying requests.
  * Various build system fixes/improvements.
  * Use Boost for signals & slots to reduce dependence on Qt's event loop.

v. 1.38.0 (2024-01-08)

  * Publish refresh action for triggering WebSocket-over-HTTP requests.
  * Ability to read signing secrets from files.
  * Move Condure into Pushpin and name the program pushpin-condure.
  * Fix crash when writing partial uncompressed WebSocket frame.
  * Fix WebSocket proxying flow control.
  * Support receiving non-chunked responses of indefinite length.
  * Qt 6 compatibility.
  * Remove configure script. Configure via environment variables instead.

v. 1.37.0 (2023-06-29)

  * Ability to use Condure instead of Zurl for outgoing connections.
  * Ability to set mode/user/group when listening on Unix socket.
  * WebSocket performance optimizations.
  * New config options: allow_compression, stats_connection_send, cdn_loop.
  * Relicense to Apache 2.0.

v. 1.36.0 (2022-11-14)

  * Ability to accept client connections via IPv6.
  * Ability to sign requests using EC or RSA public keys.
  * Include bytes/messages in report stats.

v. 1.35.0 (2022-03-11)

  * Add support for Prometheus metrics.
  * Ability to listen on a Unix socket for client connections.
  * New config options: prometheus_port, prometheus_prefix.
  * New config option: local_ports.
  * New config option: accept_pushpin_route.
  * New route condition option: no_grip.
  * Use the route of the initial request for retries and link requests.
  * pushpin-publish: fix sending hint action for http-response format.

v. 1.34.0 (2021-11-30)

  * New config option: message_wait.
  * Publish command for publishing via command socket.

v. 1.33.1 (2021-08-09)

  * Build system fixes.

v. 1.33.0 (2021-08-08)

  * Performance optimizations.
  * New config option: sig_iss.

v. 1.32.2 (2021-06-09)

  * Fix publishing to SockJS WebSocket connections.

v. 1.32.1 (2021-05-13)

  * Build system fixes.

v. 1.32.0 (2021-05-11)

  * pushpin-publish: support sending via HTTP, and do this by default.
  * pushpin-publish: support authentication.
  * pushpin-publish: use GRIP_URL environment variable if present.
  * Add Rust code to the build process.

v. 1.31.0 (2020-11-06)

  * Use Condure instead of Mongrel2, by default.
  * Ability to refresh WebSocket-over-HTTP sessions by channel.
  * Fix crash when sending delayed WebSocket messages.

v. 1.30.0 (2020-07-29)

  * Optional support for Condure instead of Mongrel2.
  * ZHTTP compatibility fixes.

v. 1.29.0 (2020-07-15)

  * Fix crash when parsing Accept header received on control port.
  * Fix crash when response hold times out while pausing.
  * Fix handling of hints in response mode.
  * Fix handling of ZeroMQ errors, including EINTR.
  * ZHTTP compatibility fixes.

v. 1.28.0 (2020-04-08)

  * New route target option: one_event.

v. 1.27.0 (2020-03-10)

  * WebSocket: ability to publish close reason.
  * WebSocket: proxy the content of ping and pong frames.

v. 1.26.0 (2019-12-11)

  * Respond with status 200 on HTTP control port root path.

v. 1.25.0 (2019-11-20)

  * Set the Mongrel2 log level and capture debug output.
  * Ability to set different log levels per subprocess.

v. 1.24.0 (2019-08-06)

  * runner: capture Mongrel2 logs when --merge-output is used.

v. 1.23.0 (2019-07-03)

  * Support log levels 0 and 1.
  * Don't write to Mongrel2 access log for log levels < 2.
  * Support JSON framing on the input PULL and SUB sockets.
  * New config option: push_in_sub_specs.
  * New config option: push_in_sub_connect.

v. 1.22.0 (2019-06-17)

  * New filter: var-subst.
  * Support content-filters field in ws-message format.

v. 1.21.0 (2019-05-01)

  * GRIP keep-alive modes: idle (default) and interval.
  * Don't put GRIP headers in Access-Control-Expose-Headers.

v. 1.20.3 (2019-04-08)

  * Fix Grip-Last values when route prefix is used.

v. 1.20.2 (2019-03-25)

  * WebSocket-Over-HTTP: fix mem leak when clients disconnect during close.

v. 1.20.1 (2019-02-20)

  * WebSocket-Over-HTTP: don't forward Content-Length header.

v. 1.20.0 (2019-02-19)

  * WebSocket-Over-HTTP: break up response messages to fit session buffers.
  * New config option: stats_format.
  * New config option: client_buffer_size.

v. 1.19.1 (2019-01-10)

  * WebSocket: fix crash when receiving frames after close frame.
  * WebSocket: include reason and headers in rejection responses.

v. 1.19.0 (2018-12-18)

  * WebSocket: support close reasons.

v. 1.18.0 (2018-08-20)

  * WebSocket-Over-HTTP: update headers (mainly Grip-Sig) for each request.
  * WebSocket-Over-HTTP: properly report errors and handle target failover.

handler/pushpin-handler.py
# Copyright (C) 2012-2013 Fan Out Networks, Inc.
#
# This file is part of Pushpin.
#
# Pushpin is free software: you can redistribute it and/or modify it under
# the terms of the GNU Affero General Public License as published by the Free
# Software Foundation, either version 3 of the License, or (at your option)
# any later version.
#
# Pushpin is distributed in the hope that it will be useful, but WITHOUT ANY
# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for
# more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program. If not, see <http://www.gnu.org/licenses/>.

import sys
import time
import threading
import ConfigParser
import json
import zmq
import tnetstring
import httpinterface
from validation import validate_publish, validate_http_publish, ValidationError
from conversion import ensure_utf8, convert_json_transport

config_file = "/etc/pushpin/pushpin.conf"
for arg in sys.argv:
	if arg.startswith("--config="):
		config_file = arg[9:]
		break

config = ConfigParser.ConfigParser()
config.read([config_file])

ctx = zmq.Context()

lock = threading.Lock()
channels = dict()

def header_remove(headers, name):
	for k in headers.keys():
		if k.lower() == name:
			del headers[k]
			return


HTTP_FORMAT = "HTTP/1.1 %(code)s %(status)s\r\n%(headers)s\r\n\r\n%(body)s"

def http_response(body, code, status, headers):
	payload = {"code": code, "status": status, "body": body}
	header_remove(headers, "content-length")
	headers["Content-Length"] = len(body)
	payload["headers"] = "\r\n".join("%s: %s" % (k, v) for k, v in
		headers.items())

	return HTTP_FORMAT % payload

def reply_http(sock, rid, code, status, headers, body):
	header = "%s %d:%s," % (rid[0], len(rid[1]), rid[1])

	if isinstance(status, unicode):
		status = status.encode("utf-8")

	# ensure headers are utf-8
	tmp = dict()
	for k, v in headers.iteritems():
		if isinstance(k, unicode):
				k = k.encode("utf-8")
		if isinstance(v, unicode):
				v = v.encode("utf-8")
		tmp[k] = v
	headers = tmp

	if isinstance(body, unicode):
		body = body.encode("utf-8")

	msg = http_response(body, code, status, headers)
	m_raw = header + " " + msg
	sock.send(m_raw)

def inspect_worker():
	sock = ctx.socket(zmq.REP)
	sock.connect(config.get("handler", "proxy_inspect_spec"))

	while True:
		m_raw = sock.recv()
		m = tnetstring.loads(m_raw)
		print "IN: %s" % m

		# reply saying to always proxy
		id = m["id"]
		#path = m["path"]
		m = dict()
		m["id"] = id
		m["no-proxy"] = False
		#m["sharing-key"] = path

		print "OUT: %s" % m
		m_raw = tnetstring.dumps(m)
		sock.send(m_raw)

	sock.close()

def accept_worker():
	sock = ctx.socket(zmq.PULL)
	sock.connect(config.get("handler", "proxy_accept_in_spec"))

	while True:
		m_raw = sock.recv()
		m = tnetstring.loads(m_raw)
		print "IN: %s" % m
		req = m["requests"][0]
		sender = req["rid"]["sender"]
		id = req["rid"]["id"]
		instruct = json.loads(m["response"]["body"])
		hold = instruct["hold"]
		channel = hold["channels"][0]["name"]
		lock.acquire()
		hchannel = channels.get(channel)
		if not hchannel:
			hchannel = set()
			channels[channel] = hchannel
		hchannel.add(((sender, id), req.get("jsonp-callback")))
		lock.release()

	sock.close()

def push_in_zmq_worker():
	in_sock = ctx.socket(zmq.PULL)
	in_sock.bind(config.get("handler", "push_in_spec"))

	out_sock = ctx.socket(zmq.PUSH)
	out_sock.connect("inproc://push_in")

	while True:
		m_raw = in_sock.recv()
		try:
			try:
				m = tnetstring.loads(m_raw)
			except:
				raise ValidationError("bad format (not a tnetstring)")

			m = validate_publish(m)

		except ValidationError as e:
			print "warning: %s, dropping" % e

		out_sock.send(tnetstring.dumps(m))

	out_sock.linger = 0
	out_sock.close()

# return None for success or string on error
def push_in_http_handler(context, channel, m):
	out_sock = context["out_sock"]

	try:
		m = validate_http_publish(m)
	except ValidationError as e:
		return e.message

	for n, i in enumerate(m["items"]):
		out = dict()

		out["channel"] = ensure_utf8(channel)

		id = i.get("id")
		if id is not None:
			out["id"] = ensure_utf8(id)

		prev_id = i.get("prev-id")
		if prev_id is not None:
			out["prev-id"] = ensure_utf8(prev_id)

		for transport in ("http-response", "http-stream"):
			if transport in i:
				out[transport] = convert_json_transport(i[transport])

		out_sock.send(tnetstring.dumps(out))

def push_in_http_worker():
	out_sock = ctx.socket(zmq.PUSH)
	out_sock.connect("inproc://push_in")

	context = dict()
	context["out_sock"] = out_sock
	httpinterface.run(config.get("handler", "push_in_http_addr"), int(config.get("handler", "push_in_http_port")), push_in_http_handler, context)

	out_sock.linger = 0
	out_sock.close()

def push_in_worker(c):
	in_sock = ctx.socket(zmq.PULL)
	in_sock.bind("inproc://push_in")
	c.acquire()
	c.notify()
	c.release()

	out_sock = ctx.socket(zmq.PUB)
	for spec in config.get("handler", "m2_out_specs").split(","):
		out_sock.connect(spec)

	while True:
		m_raw = in_sock.recv()
		m = tnetstring.loads(m_raw)
		print "IN: %s" % m
		channel = m["channel"]
		rids = set()
		lock.acquire()
		hchannel = channels.get(channel)
		if hchannel:
			reqs = hchannel
			del channels[channel]
		lock.release()
		print "relaying to %d subscribers" % len(rids)
		for req in reqs:
			rid = req[0]
			callback = req[1]

			headers = dict()
			body = m["http-response"]["body"]
			if callback is not None:
				result = dict()
				result["code"] = 200
				result["reason"] = "OK"
				result["headers"] = dict()
				result["headers"]["Content-Length"] = len(body)
				result["body"] = body

				body = callback + "(" + json.dumps(result) + ");\n"
				headers["Content-Type"] = "application/javascript"
				headers["Content-Length"] = len(body)

			reply_http(out_sock, rid, 200, "OK", headers, body)

	in_sock.close()

inspect_thread = threading.Thread(target=inspect_worker)
inspect_thread.start()

accept_thread = threading.Thread(target=accept_worker)
accept_thread.start()

# we use a condition here to ensure the inproc bind succeeds before progressing
c = threading.Condition()
c.acquire()
push_in_thread = threading.Thread(target=push_in_worker, args=(c,))
push_in_thread.start()
c.wait()
c.release()

push_in_zmq_thread = threading.Thread(target=push_in_zmq_worker)
push_in_zmq_thread.start()

push_in_http_thread = threading.Thread(target=push_in_http_worker)
push_in_http_thread.start()

try:
	while True:
		time.sleep(60)
except KeyboardInterrupt:
	pass

httpinterface.stop()
ctx.term()
